application/src/Service/ThemeManagerFactory.php

```diff
@@ -43,11 +48,35 @@ public function __invoke(ContainerInterface $serviceLocator, $requestedName, arr
                 $ini = $ini['info'];
             }
 
-            if (!$manager->iniIsValid($ini)) {
+            $theme->setIni($ini);
+            $theme->setConfigSpec($configSpec);
+
+            // Theme INI must be valid
+            if (!$manager->iniIsValid($theme)) {
+                $theme->setState(ThemeManager::STATE_INVALID_INI);
+                continue;
+            }
+
+            $omekaConstraint = $theme->getIni('omeka_version_constraint');
+            if ($omekaConstraint !== null && !Semver::satisfies(CoreModule::VERSION, $omekaConstraint)) {
+                $theme->setState(ThemeManager::STATE_INVALID_OMEKA_VERSION);
                 continue;
             }
 
-            $manager->registerTheme(new Theme($dir->getBasename(), $ini, $configSpec));
+            $theme->setState(ThemeManager::STATE_ACTIVE);
+        }
+
+        // Get all themes from the database.
+        $statement = $connection->prepare('SELECT DISTINCT(theme) FROM site');
```

Comment:
I'd prefer to see us handle the "missing theme" state on a more site-by-site basis.

Meaning, instead of checking all the assigned themes every request to see if any of them are something we don't recognize, do something more akin to our "Fallback" support elsewhere: when we're loading up a site and it requests a nonexistent theme, handle the problem then.